pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,277 @@
+				tc.workqueue.AddRateLimited(tag + key)
+				return fmt.Errorf("error tagging '%s': %s, requeuing", key, err.Error())
+			}
+		} else {
+			if err := tc.untagNodeResources(node); err != nil {
+				tc.workqueue.AddRateLimited(key)
+				return fmt.Errorf("error untagging '%s': %s, requeuing", key, err.Error())
+			}
+		}
+
+		tc.workqueue.Forget(obj)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+	}
+}
+
+// tagNodesResources tag node resources from a list of nodes
+// If we want to tag more resources, modify this function appropriately
+func (tc *TaggingController) tagNodesResources(node *v1.Node) error {
+	for _, resource := range tc.resources {
+		switch resource {
+		case opt.Instance:
+			err := tc.tagEc2Instance(node)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// tagEc2Instances applies the provided tags to each EC2 instance in
+// the cluster.
+func (tc *TaggingController) tagEc2Instance(node *v1.Node) error {
+	instanceId, err := awsv1.KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
+
+	if err != nil {
+		klog.Errorf("Error in getting instanceID for node %s, error: %v", node.GetName(), err)
+		return err
+	} else {
+		err := tc.cloud.TagResource(string(instanceId), tc.tags)
+
+		if err != nil {
+			klog.Errorf("Error in tagging EC2 instance for node %s, error: %v", node.GetName(), err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// untagNodeResources untag node resources from a list of nodes
+// If we want to untag more resources, modify this function appropriately
+func (tc *TaggingController) untagNodeResources(node *v1.Node) error {
+	for _, resource := range tc.resources {
+		switch resource {
+		case opt.Instance:
+			err := tc.untagEc2Instance(node)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// untagEc2Instances deletes the provided tags to each EC2 instances in
+// the cluster.
+func (tc *TaggingController) untagEc2Instance(node *v1.Node) error {
+	instanceId, err := awsv1.KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
+
+	if err != nil {
+		klog.Errorf("Error in getting instanceID for node %s, error: %v", node.GetName(), err)
+		return err
+	} else {
+		err := tc.cloud.UntagResource(string(instanceId), tc.tags)
+
+		if err != nil {
+			klog.Errorf("Error in untagging EC2 instance for node %s, error: %v", node.GetName(), err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// enqueueNode takes in the object to enqueue to the workqueue and whether
+// the object is to be tagged
+func (tc *TaggingController) enqueueNode(obj interface{}, isTagged bool) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	if isTagged {
+		tc.workqueue.Add(tag + key)
+	} else {
+		tc.workqueue.Add(key)
+	}
+}
+
+func (tc *TaggingController) getActionAndKey(key string) (bool, string) {
+	isTagged := false
```

Comment:
So does `isTagged` mean it is already tagged or it needs to be tagged?  If the latter, rename the variable.

Reply:
renamed the variable